src/test/java/rx/ObservableTests.java

```diff
@@ -1174,4 +1175,304 @@ public Object call(OnSubscribe<Object> onSubscribe) {
+        Action0 onCompleted = Actions.empty();
+        try {
+            Observable.just(1).forEach((Action1<Integer>)null);
+            fail("Should have thrown IllegalArgumentException");
+        } catch (IllegalArgumentException ex) {
+            assertEquals("onNext can not be null", ex.getMessage());
+        }
+
+        try {
+            Observable.just(1).forEach(null, onError);
+            fail("Should have thrown IllegalArgumentException");
+        } catch (IllegalArgumentException ex) {
+            assertEquals("onNext can not be null", ex.getMessage());
+        }
+
+        try {
+            Observable.just(1).forEach(null, onError, onCompleted);
+            fail("Should have thrown IllegalArgumentException");
+        } catch (IllegalArgumentException ex) {
+            assertEquals("onNext can not be null", ex.getMessage());
+        }
+        
+        try {
+            Observable.just(1).forEach(onNext, null);
+            fail("Should have thrown IllegalArgumentException");
+        } catch (IllegalArgumentException ex) {
+            assertEquals("onError can not be null", ex.getMessage());
+        }
+
+        try {
+            Observable.just(1).forEach(onNext, null, onCompleted);
+            fail("Should have thrown IllegalArgumentException");
+        } catch (IllegalArgumentException ex) {
+            assertEquals("onError can not be null", ex.getMessage());
+        }
+        
+        try {
+            Observable.just(1).forEach(onNext, onError, null);
+            fail("Should have thrown IllegalArgumentException");
+        } catch (IllegalArgumentException ex) {
+            assertEquals("onComplete can not be null", ex.getMessage());
+        }
+    }
+    
+    @Test
+    public void observableThrowsWhileSubscriberIsUnsubscribed() {
+        TestSubscriber<Object> ts = TestSubscriber.create();
+        ts.unsubscribe();
+        
+        final List<Throwable> list = new ArrayList<Throwable>();
+        
+        RxJavaHooks.setOnError(new Action1<Throwable>() {
+            @Override
+            public void call(Throwable t) {
+                list.add(t);
+            }
+        });
+        
+        try {
+            new FailingObservable().subscribe(ts);
+            
+            assertEquals(1, list.size());
+            
+            assertEquals("Forced failure", list.get(0).getMessage());
+        } finally {
+            RxJavaHooks.reset();
+        }
+    }
+    
+    @Test
+    public void observableThrowsWhileOnErrorFails() {
+        Subscriber<Object> ts = new SafeSubscriber<Object>(new TestSubscriber<Object>()) {
+            @Override
+            public void onError(Throwable e) {
+                throw new TestException();
+            }
+        };
+        
+        try {
+            new FailingObservable().subscribe(ts);
+            fail("Should have thrown OnErrorFailedException");
+        } catch (OnErrorFailedException ex) {
+            // expected
```

Comment:
If you really wanted to be comprehensive here you could put `TestException` in a method local and verify `assertSame(ex.getCause(), testException)`.